# LIVE CODE 3 

NAME   : AAN NURLIYANAH

BATCH  : 002

## Problem


Kamu adalah seorang data analis di Austin Smart City. Saat ini, kamu sedang membantu kepolisian kota Austin untuk mempelajari kondisi tindak kejahatan di kota Austin.

Dataset yang akan digunakan dapat diakses menggunakan bigquery-public-data pada Google Cloud Big Query.
Data yang digunakan adalah tabel crime pada dataset austin_crime.

## Working Area

### Library 

Library yang digunakan adalah pandas

In [1]:
import pandas as pd


Mengkoneksikan `BigQuery` ke Google Colab tempat mengerjakan live code 3 dengan code berikut:

from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

project_id = "aan-nurliyanah" 
client = bigquery.Client(project=project_id)

untuk project name di GCP saya dengan nama `aan-nurliyanah`. Setelah dijalankan query akan meminta akses ke GCP, lanjutkan hingga result nya 'Authenticated'.

In [2]:
from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

project_id = "aan-nurliyanah" 
client = bigquery.Client(project=project_id)

Authenticated


Memanggil dataset yang akan digunakan untuk kita analisis data di dalamnya secara garis besar. tabel yang akan digunakan yaitu tabel crime. 

In [3]:
tabel_crime = client.query('''
SELECT * FROM `bigquery-public-data.austin_crime.crime`
''').to_dataframe()

tabel_crime

unique_key                                            address  \
0       2.016500e+10                                         Austin, TX   
1       2.016500e+10                                         Austin, TX   
2       2.016251e+09                                         Austin, TX   
3       2.016500e+10                                         Austin, TX   
4       2.016322e+09                                         Austin, TX   
...              ...                                                ...   
116670  2.014800e+10  13813 N MOPAC EXPY NB Austin, TX 78727 (30.439...   
116671  2.014502e+10  14205 N MOPAC EXPY NB Austin, TX 78727 (30.444...   
116672  2.014503e+10  12761 COUNCIL BLUFF DR Austin, TX 78727 (30.42...   
116673  2.014503e+10  12805 COUNCIL BLUFF DR Austin, TX 78727 (30.42...   
116674  2.014506e+10  13305 COUNCIL BLUFF DR Austin, TX 78727 (30.43...   

        census_tract            clearance_date clearance_status  \
0                NaN 2016-03-14 12:00:00+00:00      Not cleared   
1                NaN 2016-03-31 12:00:00+00:00      Not cleared   
2                NaN 2016-09-16 12:00:00+00:00      Not cleared   
3                NaN 2016-02-09 12:00:00+00:00      Not cleared   
4                NaN 2016-03-22 12:00:00+00:00      Not cleared   
...              ...                       ...              ...   
116670         18.46 2014-03-05 12:00:00+00:00      Not cleared   
116671         18.46 2014-06-02 12:00:00+00:00      Not cleared   
116672         18.46 2014-07-31 12:00:00+00:00      Not cleared   
116673         18.46 2014-07-24 12:00:00+00:00      Not cleared   
116674         18.46 2015-01-02 12:00:00+00:00      Not cleared   

        council_district_code                     description district  \
0                         NaN  RAPE                                 UK   
1                         NaN  RAPE                                 UK   
2                         NaN  RAPE                                 UK   
3                         NaN  AGG RAPE OF A CHILD                  UK   
4                         NaN  RAPE OF A CHILD                      UK   
...                       ...                             ...      ...   
116670                    7.0  BURGLARY OF VEHICLE                   A   
116671                    7.0  THEFT                                 A   
116672                    7.0  THEFT                                 A   
116673                    7.0  THEFT OF AUTO PARTS                   A   
116674                    7.0  THEFT                                 A   

         latitude  longitude                location    location_description  \
0             NaN        NaN                    None                     nan   
1             NaN        NaN                    None                     nan   
2             NaN        NaN                    None                     nan   
3             NaN        NaN                    None                     nan   
4             NaN        NaN                    None                     nan   
...           ...        ...                     ...                     ...   
116670  30.439137 -97.698275  (30.439137,-97.698275)   13813 N MOPAC EXPY NB   
116671  30.444353 -97.696183  (30.444353,-97.696183)   14205 N MOPAC EXPY NB   
116672  30.429571 -97.717999  (30.429571,-97.717999)  12761 COUNCIL BLUFF DR   
116673  30.429703 -97.718265  (30.429703,-97.718265)  12805 COUNCIL BLUFF DR   
116674  30.435655 -97.715325  (30.435655,-97.715325)  13305 COUNCIL BLUFF DR   

                    primary_type                 timestamp  x_coordinate  \
0                           Rape 2016-01-19 12:00:00+00:00           NaN   
1                           Rape 2016-01-25 12:00:00+00:00           NaN   
2                           Rape 2016-01-25 12:00:00+00:00           NaN   
3                           Rape 2016-01-27 12:00:00+00:00           NaN   
4                           Rape 2016-02-01 12:00:00+00:00           NaN  

## PROBLEM STATEMENT

Definisikan problem statement berdasarkan penjabaran masalah di bawah yang akan teman-teman jawab menggunakan SQL!


Jawab :

Meningkatkan penyelesaian kasus pencurian yang belum terselesaikan di kota Austin di district D pada awal 2017 berdasarkan data pada tahun 2016. 

## Penjabaran Masalah


Soal 1 : Tampilkan tabel crime sampai 5 baris pertama

In [20]:
tabel_crime = client.query('''
SELECT * FROM `bigquery-public-data.austin_crime.crime`
''').to_dataframe()

tabel_crime.head(5)

unique_key     address  census_tract            clearance_date  \
0  2.016500e+10  Austin, TX           NaN 2016-03-14 12:00:00+00:00   
1  2.016500e+10  Austin, TX           NaN 2016-03-31 12:00:00+00:00   
2  2.016251e+09  Austin, TX           NaN 2016-09-16 12:00:00+00:00   
3  2.016500e+10  Austin, TX           NaN 2016-02-09 12:00:00+00:00   
4  2.016322e+09  Austin, TX           NaN 2016-03-22 12:00:00+00:00   

  clearance_status  council_district_code                     description  \
0      Not cleared                    NaN  RAPE                             
1      Not cleared                    NaN  RAPE                             
2      Not cleared                    NaN  RAPE                             
3      Not cleared                    NaN  AGG RAPE OF A CHILD              
4      Not cleared                    NaN  RAPE OF A CHILD                  

  district  latitude  longitude location location_description primary_type  \
0       UK       NaN        NaN     None                  nan         Rape   
1       UK       NaN        NaN     None                  nan         Rape   
2       UK       NaN        NaN     None                  nan         Rape   
3       UK       NaN        NaN     None                  nan         Rape   
4       UK       NaN        NaN     None                  nan         Rape   

                  timestamp  x_coordinate  y_coordinate  year zipcode  
0 2016-01-19 12:00:00+00:00           NaN           NaN  2016    None  
1 2016-01-25 12:00:00+00:00           NaN           NaN  2016    None  
2 2016-01-25 12:00:00+00:00           NaN           NaN  2016    None  
3 2016-01-27 12:00:00+00:00           NaN           NaN  2016    None  
4 2016-02-01 12:00:00+00:00           NaN           NaN  2016    None

Untuk menampilkan tabel crime sampai 5 baris pertama menggunakan fungsi head(5). 

Soal 2 : Apa saja jenis tindak kejahatan yang terjadi di kota Austin selama tahun 2016?

Untuk mencari jenis tindak kejahatan apa saja menggunakan kolom primary_type dan description. Kemudian di cari berdasarkan klausa where dimana data yang diminta hanya tahun 2016 saja diambil dari kolom year, data diurutkan secara ascending dari berdasarkan primary_type.

In [13]:
jenis_kejahatan = client.query('''
SELECT primary_type, description
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016
ORDER BY primary_type ASC
''').to_dataframe()

jenis_kejahatan

primary_type                     description
0      Agg Assault  AGG ASLT STRANGLE/SUFFOCATE   
1      Agg Assault  AGG ASSAULT FAM/DATE VIOLENCE 
2      Agg Assault  AGG ASSAULT                   
3      Agg Assault  AGG ASSAULT                   
4      Agg Assault  AGG ASSAULT                   
...            ...                             ...
37456        Theft  BURGLARY OF VEHICLE           
37457        Theft  THEFT OF LICENSE PLATE        
37458        Theft  THEFT FROM BUILDING           
37459        Theft  BURGLARY OF VEHICLE           
37460        Theft  THEFT                         

[37461 rows x 2 columns]

Dari hasil pencarian diatas didaptkan jenis jenis type kejahatan beserta deskripsinya diantaranya ada Agg Assault, dan Theft. 

Soal 3 : Berapa banyak kasus tindak kejahatan berdasarkan kasus pembersihannya (clearance status) selama tahun 2016?

Untuk mencari banyaknya kasus tindak kejahatan kita cari menggunakan fungsi count, yang dihitung adalah dari deskripsinya kemudian dicari menggunakan klausa where dimana data yg diminta di tahun 2016 saja di kolom year, kemudian data di group by berdasarkan clearance statusnya. Untuk total case diurutkan secara descending dari besar ke kecil.

In [9]:
kasus_kejahatan = client.query('''
SELECT 
  COUNT (description) AS total_case,
  clearance_status
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016
GROUP BY clearance_status
ORDER BY total_case DESC
''').to_dataframe()

kasus_kejahatan

total_case      clearance_status
0       29715           Not cleared
1        5013     Cleared by Arrest
2        1719                  None
3        1014  Cleared by Exception

Dari perhitungan diatas didapatkan total case dengan cleareance status 'not cleared' adalah yang tertinggi yaitu 29715, artinya banyak kasus yang tidak terpecahkan di tahun 2016 ini. Jumlah ini sangat ekstreme dibandingkan dengan kasus yang sudah cleared/ terselesaikan. 

Saol 4 : Jenis kejahatan apa saja yang paling banyak belum terselesaikan kasusnya di tahun 2016?

Untuk mencari jenis kejahatan yang paling banyak belum terselesaikan dicari dengan menggunakan fungsi count untuk menghitung total case, dihitung berdasarkan deskripsinya, kemudian di group by berdasarkan primary_type, description dan clearence status. Dimana data yang dicari adalah di tahun 2016 dan status celaresnce = 'not cleared'. Data diurutkan berdasarkan total case dari yang terbesar ke kecil.

In [12]:
kasus_belum_selesai = client.query('''
SELECT 
  primary_type, description , clearance_status,
  COUNT (description) AS total_case,
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016 AND 
  clearance_status = 'Not cleared'
GROUP BY primary_type, description, clearance_status
ORDER BY total_case DESC

''').to_dataframe()

kasus_belum_selesai

primary_type                     description clearance_status  total_case
0         Theft  THEFT                               Not cleared        8607
1         Theft  BURGLARY OF VEHICLE                 Not cleared        8089
2      Burglary  BURGLARY OF RESIDENCE               Not cleared        2706
3         Theft  THEFT BY SHOPLIFTING                Not cleared        1565
4    Auto Theft  AUTO THEFT                          Not cleared        1540
5      Burglary  BURGLARY NON RESIDENCE              Not cleared        1441
6         Theft  THEFT OF BICYCLE                    Not cleared        1346
7         Theft  THEFT FROM AUTO                     Not cleared         633
8   Agg Assault  AGG ASSAULT                         Not cleared         530
9         Theft  THEFT OF LICENSE PLATE              Not cleared         445
10        Theft  THEFT FROM PERSON                   Not cleared         362
11      Robbery  AGG ROBBERY/DEADLY WEAPON           Not cleared         298
12     Burglary  BURG NON RESIDENCE SHEDS            Not cleared         287
13        Theft  THEFT OF AUTO PARTS                 Not cleared         283
14        Theft  THEFT FROM BUILDING                 Not cleared         241
15      Robbery  ROBBERY BY ASSAULT                  Not cleared         210
16        Theft  THEFT OF TRAILER                    Not cleared         151
17  Agg Assault  AGG ASLT STRANGLE/SUFFOCATE         Not cleared         144
18         Rape  RAPE                                Not cleared         128
19  Agg Assault  DEADLY CONDUCT                      Not cleared         114
20        Theft  BREACH OF COMPUTER SECURITY         Not cleared          87
21  Agg Assault  AGG ASSAULT FAM/DATE VIOLENCE       Not cleared          70
22         Rape  SEXUAL ASSAULT W/ OBJECT            Not cleared          64
23         Rape  SEXUAL ASSAULT OF CHILD/OBJECT      Not cleared          57
24         Rape  RAPE OF A CHILD                     Not cleared          52
25      Robbery  ROBBERY BY THREAT                   Not cleared          44
26  Agg Assault  AGG ASSAULT WITH MOTOR VEH          Not cleared          39
27        Theft  THEFT OF METAL                      Not cleared          36
28        Theft  BURGLARY OF COIN-OP MACHINE         Not cleared          25
29         Rape  FORCED SODOMY                       Not cleared          20
30         Rape  AGG RAPE OF A CHILD                 Not cleared          12
31         Rape  AGG SEXUAL ASSAULT CHILD/OBJEC      Not cleared          12
32        Theft  THEFT OF HEAVY EQUIPMENT            Not cleared          11
33  Agg Assault  AGG ASLT W/MOTOR VEH FAM/DAT V      Not cleared           8
34      Robbery  AGG ROBBERY BY ASSAULT              Not cleared           8
35         Rape  AGG RAPE                            Not cleared           7
36  Agg Assault  AGG ASLT ENHANC STRANGL/SUFFOC      Not cleared           7
37     Burglary  BURG OF RES - SEXUAL NATURE         Not cleared           6
38  Agg Assault  AGG ASSAULT ON PUBLIC SERVANT       Not cleared           6
39         Rape  AGG SEXUAL ASSAULT W OBJECT         Not cleared           5
40        Theft  PURSE SNATCHING                     Not cleared           4
41        Theft  POCKET PICKING                      Not cleared           3
42        Theft  MISAPPLY FIDUCIARY PROP             Not cleared           3
43        Theft  THEFT/TILL TAPPING                  Not cleared           2
44         Rape  AGG FORCED SODOMY                   Not cleared           2
45        Theft  THEFT CATALYTIC CONVERTER           Not cleared           1
46  Agg Assault  DEADLY CONDUCT FAM/DATE VIOL        Not cleared           1
47         Rape  AGG FORCED SODOMY OF CHILD          Not cleared           1
48       Murder  MURDER                              Not cleared           1
49         Rape  FORCED SODOMY OF CHILD              Not cleared           1

Dari perhitungan diatas didapatkan hasil yaitu paling banyak total 8607 kasus yang tidak terselesaikan yaitu ada pada jenis kejahatan pencurian (theft). 

Soal 5 : Di bulan apa di tahun 2016 kasus pencurian (theft) sangat banyak terjadi?

Untuk mencari bulan apa di tahun 2016 kasus theft banyak terjadi menggunakan subquery dimana pertama kita mencari data perbulan dahulu dengan menggunakan fungsi extraxt pada timestamp dimana kejadian pencurian dimulai, kemudian di hitung total case nya dengan menggunakan fungsi count, dengan kondisi tahun yg dicari 2016 dan jenis kejahatan = theft. Setelah didapatkan data setiap bulannya kemudian di urutkan dengan menggunakan fungsi row_number dan di partisi per bulan untuk mencari urutan pertama saja yang ditampilkan perbulannya. 

In [11]:
kasus_bulan = client.query('''
SELECT
crime_type, description, month, total_case
FROM (SELECT *, row_number() over (
PARTITION by month 
ORDER BY total_case DESC) AS URUTAN
FROM 
(SELECT 
  primary_type as crime_type,
  description as description,
  EXTRACT (month from timestamp) as month,
  COUNT (description) AS total_case,
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016 AND 
  primary_type = 'Theft'
GROUP BY month, crime_type, description
ORDER BY total_case DESC))
WHERE URUTAN = 1
ORDER BY total_case DESC
''').to_dataframe()

kasus_bulan

crime_type                     description  month  total_case
0       Theft  THEFT                              12         925
1       Theft  THEFT                               4         900
2       Theft  THEFT                               6         898
3       Theft  THEFT                               3         884
4       Theft  THEFT                              10         882
5       Theft  THEFT                              11         881
6       Theft  THEFT                               9         874
7       Theft  THEFT                               1         872
8       Theft  THEFT                               7         864
9       Theft  THEFT                               8         858
10      Theft  THEFT                               5         852
11      Theft  THEFT                               2         752

Dari perhitungan di atas didapatkan hasil di bulan 12 atau bulan desember terjadi kasus pencurian terbanyak yaitu  925 kasus. 

Soal 6 : Di distrik apa yang paling banyak terjadi kasus pencurian selama 2016?

Untuk mencari district menggunakan metode count di bagian description untuk mencari total case kemudian di group by berdasarkan district .

In [15]:
kasus_district = client.query('''
SELECT 
primary_type as crime_type,
description as description,
district as district,
COUNT (description) AS total_case,
FROM `bigquery-public-data.austin_crime.crime` 
WHERE year = 2016 AND 
primary_type = 'Theft'
GROUP BY crime_type, description, district
ORDER BY total_case DESC
''').to_dataframe()

kasus_district

crime_type                     description district  total_case
0        Theft  THEFT                                 D        1597
1        Theft  BURGLARY OF VEHICLE                   D        1344
2        Theft  THEFT                                 F        1230
3        Theft  THEFT                                 I        1140
4        Theft  THEFT                                 B        1131
..         ...                             ...      ...         ...
149      Theft  PURSE SNATCHING                       B           1
150      Theft  THEFT/TILL TAPPING                    F           1
151      Theft  THEFT OF HEAVY EQUIPMENT              F           1
152      Theft  PURSE SNATCHING                       H           1
153      Theft  THEFT OF HEAVY EQUIPMENT              I           1

[154 rows x 4 columns]

dari hasil perhitungan diatas didaptakan diatrict yang terjadi banyak kejahatan adalah di district D dengan total 1597 kasus.

Soal 7 : Berapa lama rata-rata waktu (dalam hari) kasus tindak kejahatan 'theft' terselesaikan sejak kasus tersebut terangkat selama tahun 2016? (Hint: Hitung selisih hari dari timestamp hingga clearance date)

Untuk menghitung rata rata waktu menggunakan fungsi timestamp_diff dimana kita mencari selisih waktu dalam hari antara clearence date dan timestamp. 
sebelum mencari rata rata saya menampilkan dulu jangka waktu / selisih waktunya secara rinci diurutkan dari besar ke kecil. 

In [17]:
sebelum_rata_rata = client.query('''
SELECT
	clearance_status ,timestamp, clearance_date,
    timestamp_diff(clearance_date,timestamp,day) as Jangka_waktu_hari
 FROM bigquery-public-data.austin_crime.crime
 WHERE year = 2016 AND
 primary_type = 'Theft' AND 
 clearance_status = 'Cleared by Arrest' OR 
 clearance_status = 'Cleared by Exception'
 ORDER BY Jangka_waktu_hari DESC
''').to_dataframe()
sebelum_rata_rata

clearance_status                 timestamp  \
0        Cleared by Arrest 2016-01-27 12:00:00+00:00   
1     Cleared by Exception 2014-05-26 12:00:00+00:00   
2        Cleared by Arrest 2016-02-17 12:00:00+00:00   
3     Cleared by Exception 2014-03-26 12:00:00+00:00   
4     Cleared by Exception 2015-03-09 12:00:00+00:00   
...                    ...                       ...   
5906     Cleared by Arrest 2016-10-27 12:00:00+00:00   
5907     Cleared by Arrest 2016-11-06 12:00:00+00:00   
5908     Cleared by Arrest 2016-11-20 12:00:00+00:00   
5909     Cleared by Arrest 2016-11-22 12:00:00+00:00   
5910     Cleared by Arrest 2016-12-16 12:00:00+00:00   

                clearance_date  Jangka_waktu_hari  
0    2017-01-20 12:00:00+00:00                359  
1    2015-05-15 12:00:00+00:00                354  
2    2017-01-23 12:00:00+00:00                341  
3    2015-02-23 12:00:00+00:00                334  
4    2016-01-25 12:00:00+00:00                322  
...                        ...                ...  
5906 2016-10-27 12:00:00+00:00                  0  
5907 2016-11-06 12:00:00+00:00                  0  
5908 2016-11-20 12:00:00+00:00                  0  
5909 2016-11-22 12:00:00+00:00                  0  
5910 2016-12-16 12:00:00+00:00                  0  

[5911 rows x 4 columns]

Setelah mendapatkan hasil, kita lanjuta subquery untuk mencari rata rata dengan fungsi avg. 

In [19]:
rata_rata = client.query('''
SELECT
AVG(Jangka_waktu_hari) AS rata_rata,
FROM(SELECT
	clearance_status ,timestamp, clearance_date,
    timestamp_diff(clearance_date,timestamp,day) as Jangka_waktu_hari
 FROM bigquery-public-data.austin_crime.crime
 WHERE year = 2016 AND
 primary_type = 'Theft' AND 
 clearance_status = 'Cleared by Arrest' OR 
 clearance_status = 'Cleared by Exception'
 ORDER BY Jangka_waktu_hari DESC)
''').to_dataframe()

rata_rata


rata_rata
0   25.70225

dari hasil perhitungan diatas didapatkan rata rata lamanya kasus terselesaikan adalah 25 hari . Untuk yang paling lama adalah 359 hari yang berakhir dengan cleared by arrest. 

### Poin Analisis


Dari penjabaran yang sudah kamu bahas (pertanyaan 2-7), apa kesimpulan yang kamu dapat sampaikan?


Jawab :

- Untuk jenis jenis type kejahatan beserta deskripsinya diantaranya ada Agg Assault, dan Theft. 

- Untuk total case dengan cleareance status 'not cleared' adalah yang tertinggi yaitu 29715, artinya banyak kasus yang tidak terpecahkan di tahun 2016 ini. Jumlah ini sangat ekstreme dibandingkan dengan kasus yang sudah cleared/ terselesaikan. 

- Paling banyak total 8607 kasus yang tidak terselesaikan yaitu ada pada jenis kejahatan pencurian (theft). 

- Di bulan 12 atau bulan desember terjadi kasus pencurian terbanyak yaitu  925 kasus. 

- district yang terjadi banyak kejahatan adalah di district D dengan total 1597 kasus.

- rata rata lamanya kasus terselesaikan adalah 25 hari . Untuk yang paling lama adalah 359 hari yang berakhir dengan cleared by arrest.

Masukan untuk kota Austin adalah sesegera mungkin menyelesaikan kasus yang belum terpecahkan. 